backstory/src/test/java/com/google/sps/servlets/BlobstoreManagerTest.java

```diff
@@ -0,0 +1,105 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.servlets;
+
+import com.google.sps.servlets.data.BlobstoreManager;
+import com.google.sps.servlets.data.BlobstoreServiceConstantFields;
+import com.google.appengine.api.blobstore.BlobInfo;
+import com.google.appengine.api.blobstore.BlobInfoFactory;
+import com.google.appengine.api.blobstore.BlobstoreService;
+import org.junit.Assert;
+import org.junit.Test;
+import java.util.ArrayList;
+import java.util.List;
+import java.io.IOException;
+import static org.mockito.Mockito.*;
+import com.google.appengine.api.blobstore.BlobKey;
+import java.util.Map;
+import java.util.HashMap;
+import java.util.List;
+import java.util.ArrayList;
+import java.io.IOException;
+import javax.servlet.http.HttpServletRequest;
+import org.junit.runner.RunWith;
+import org.mockito.runners.MockitoJUnitRunner;
+import org.junit.Before;
+
+/**
+ * Tests for the blobstore service wrapper, BlobstoreManager
+ */
+@RunWith(MockitoJUnitRunner.class)
+public final class BlobstoreManagerTest {
+  // All necessary mocks and objects for the test calls themselves
+  private BlobstoreService mockBlobstoreService;
+  private BlobstoreServiceConstantFields mockBlobstoreServiceConstantFields;
+  private BlobInfoFactory mockBlobInfoFactory;
+  private HttpServletRequest mockRequest;
+  private String formInputElementName;
+  private String mockBlobkey;
+
+  /**
+   * Sets up the shared mocks and objects for the two tests.
+   */
+  @Before
+  public void setUp() {
+    mockRequest = mock(HttpServletRequest.class);
+
+    mockBlobstoreService = mock(BlobstoreService.class);
+    mockBlobstoreServiceConstantFields = mock(BlobstoreServiceConstantFields.class);
+    mockBlobInfoFactory = mock(BlobInfoFactory.class);
+    formInputElementName = "image-upload";
+    mockBlobkey = "mockBlobkey";
+
+    BlobInfo mockBlobInfo = mock(BlobInfo.class);
+    Map<String, List<BlobKey>> mockInputToBlobkey = new HashMap<>();
+    List<BlobKey> mockBlobKeys = new ArrayList<>();
+    mockBlobKeys.add(new BlobKey(mockBlobkey));
+    mockInputToBlobkey.put(formInputElementName, mockBlobKeys);
+
+    when(mockBlobstoreService.getUploads(any(HttpServletRequest.class)))
+        .thenReturn(mockInputToBlobkey);
+
+    // Mock one image getting uploaded
+    when(mockBlobInfo.getSize()).thenReturn(new Long(1));
+    when(mockBlobInfoFactory.loadBlobInfo(any(BlobKey.class)))
+        .thenReturn(mockBlobInfo);
+  }
+
+  /**
+   * Tests the getUploadedFileBlobKeyString method of BlobstoreManager
+   */
+  @Test
+  public void testGetUploadedFileBlobKeyString() throws IOException {
+    BlobstoreManager blobstoreManager = new BlobstoreManager(mockBlobstoreService, mockBlobstoreServiceConstantFields, mockBlobInfoFactory);
+    Assert.assertEquals(mockBlobkey, blobstoreManager.getUploadedFileBlobKeyString(mockRequest, formInputElementName));
```

Comment:
Ideally the reader of a test can understand what's being tested without needing to navigate to other parts of the file. One option to make this clearer: moving relevant code from setUp to this method (currently I'm not sure what code from setUp is relevant). If needed you can set up helper methods to construct objects as long as the helper methods don't mask what is being mocked... however, in tests, duplication is considered okay, when clarity is improved (i.e. avoid setUp code that configures instances away from default values).

https://drive.google.com/file/d/15GGGANZoaG9JIyGyrTkZWkmYlaih01MJ/view

This also applies to the test case below.

Reply:
Okay sounds good, I updated the test code so that the mocks are set up in each individual method.